### PR TITLE
prefer tickers to time.Sleep

### PR DESCRIPTION
**Pull request by melekes**
Refs #790

### CODE REVIEW

**Comment by melekes** [comment]
```
BEFORE:
real    2m37.473s
user    3m5.428s
sys     0m27.388s

AFTER:
real    2m35.850s
user    3m11.070s
sys     0m27.979s
```

File: consensus/mempool_test.go [hunk 1]
```diff
@@ -134,28 +134,24 @@ func TestRmBadTx(t *testing.T) {
...
 		// check for the tx
-		for {
-			time.Sleep(time.Second)
-			txs := cs.mempool.Reap(1)
-			if len(txs) == 0 {
-				ch <- struct{}{}
-				return
-			}
-
+		txs := cs.mempool.Reap(1)
```
Comment (ebuchman) [inline consensus/mempool_test.go]: how can we remove the for loop? what if the tx hasn't been removed yet

Comment (ebuchman) [reply consensus/mempool_test.go]: would be best if we show the mempool is not empty first, and then it becomes empty when the abci app returns.

File: node/node_test.go [hunk 2]
```diff
@@ -20,19 +20,23 @@ func TestNodeStartStop(t *testing.T) {
...
 	t.Logf("Started node %v", n.sw.NodeInfo())
 
-	// Wait a bit to initialize
-	// TODO remove time.Sleep(), make asynchronous.
-	time.Sleep(time.Second * 2)
+	ticker := time.NewTicker(10 * time.Millisecond)
+	select {
+	case <-ticker.C:
+		if n.IsRunning() {
+			return
```
Comment (ebuchman) [inline node/node_test.go]: doesnt this abort the test? maybe you mean `break`?

Comment (ebuchman) [reply node/node_test.go]: Also `IsRunning` should be true right away. Maybe we should wait for a block to be produced?

File: p2p/pex_reactor_test.go [hunk 3]
```diff
@@ -98,22 +99,39 @@ func TestPEXReactorRunning(t *testing.T) {
...
 
 	// stop them
 	for _, s := range switches {
 		s.Stop()
 	}
 }
 
+func assertSomePeersWithTimeout(t *testing.T, switches []*Switch, checkPeriod, timeout time.Duration) {
+	ticker := time.NewTicker(checkPeriod)
+	select {
```
Comment (ebuchman) [inline p2p/pex_reactor_test.go]: needs a loop?

File: p2p/pex_reactor_test.go [hunk 4]
```diff
@@ -98,22 +99,39 @@ func TestPEXReactorRunning(t *testing.T) {
...
+	case <-ticker.C:
+		// check peers are connected
+		allGood := true
+		for _, s := range switches {
+			outbound, inbound, _ := s.NumPeers()
+			if outbound+inbound == 0 {
+				allGood = false
```
Comment (ebuchman) [inline p2p/pex_reactor_test.go]: can drop allGood and just continue

File: p2p/switch_test.go [hunk 5]
```diff
@@ -131,41 +131,31 @@ func TestSwitches(t *testing.T) {
...
 
-	// Check message on ch2
-	ch2Msgs := s2.Reactor("bar").(*TestReactor).getMsgs(byte(0x02))
-	if len(ch2Msgs) != 1 {
-		t.Errorf("Expected to have received 1 message in ch2")
-	}
-	if !bytes.Equal(ch2Msgs[0].Bytes, wire.BinaryBytes(ch2Msg)) {
-		t.Errorf("Unexpected message bytes. Wanted: %X, Got: %X", wire.BinaryBytes(ch2Msg), ch2Msgs[0].Bytes)
+func assertMsgReceivedWithTimeout(t *testing.T, msg string, channel byte, reactor *TestReactor, checkPeriod, timeout time.Duration) {
+	ticker := time.NewTicker(checkPeriod)
```
Comment (ebuchman) [inline p2p/switch_test.go]: loop ...

File: consensus/mempool_test.go (hunk 1 above)
Comment (melekes) [reply consensus/mempool_test.go]: looks like mempool is always empty

**Pull request merged**